src/GeniusYield/TxBuilder/Node.hs

```diff
@@ -277,12 +279,21 @@ runGYTxMonadNodeCore ownUtxoUpdateF cstrat nid providers addrs change collateral
     pp          <- gyGetProtocolParameters providers
     ps          <- gyGetStakePools providers
 
-    e <- unGYTxMonadNode (buildTxCore ss eh pp ps cstrat ownUtxoUpdateF addrs change collateral action) GYTxNodeEnv
+    collateral' <-
+      runMaybeT $ do
+        (collateralRef, toCheck) <- MaybeT $ return collateral
+        if not toCheck then return collateralRef
+        else do
+          collateralUtxo <- liftIO $ runGYTxQueryMonadNode nid providers $ utxoAtTxOutRef' collateralRef
+          if utxoValue collateralUtxo == collateralValue then return collateralRef
+          else MaybeT $ return Nothing
```

Comment:
I think it will be much cleaner if we define a `getCollateral` function locally like
```
  where
      getCollateral :: IO (Maybe GYTxOutRef)
      getCollateral = runMaybeT $ do
          (collateralRef, toCheck) <- MaybeT $ return collateral
          if not toCheck then return collateralRef
          else do
              collateralUtxo <- liftIO $ runGYTxQueryMonadNode nid providers
                                $ utxoAtTxOutRef' collateralRef
              if utxoValue collateralUtxo == collateralValue
              then return collateralRef
              else MaybeT $ return Nothing
```
And then just `collateral' <- getCollateral`. I also tried to use `hoistMaybe` instead of `MaybeT . return`, but I guess we are using an "old" transformers lib, so there is no problem with this.

Reply:
Sure.
>  I also tried to use `hoistMaybe` instead of `MaybeT . return`, but I guess we are using an "old" transformers lib, so there is no problem with this.
Yeah, would spend time to see if it can be sorted, relevant with [this](https://github.com/geniusyield/atlas/issues/105) as well.

Reply:
We can fix for it by doing something like `allow-newer: insert-ordered-containers:transformers` in our `cabal.project` file but I'll have do for it when seeing https://github.com/geniusyield/atlas/issues/101.